addons/website_sale_delivery/models/sale_order.py

```diff
@@ -124,66 +86,60 @@ def _get_delivery_methods(self, cr, uid, order, context=None):
+    @api.multi
+    def _cart_update(self, product_id=None, line_id=None, add_qty=0, set_qty=0, **kwargs):
         """ Override to update carrier quotation if quantity changed """
 
-        values = super(SaleOrder, self)._cart_update(
-            cr, uid, ids, product_id, line_id, add_qty, set_qty, context, **kwargs)
+        values = super(SaleOrder, self)._cart_update(product_id, line_id, add_qty, set_qty, **kwargs)
 
         if add_qty or set_qty is not None:
-            for sale_order in self.browse(cr, uid, ids, context=context):
-                self._check_carrier_quotation(cr, uid, sale_order, context=context)
+            for sale_order in self:
+                self._check_carrier_quotation()
 
         return values
 
-    def _get_shipping_country(self, cr, uid, values, context=None):
-        country_ids = set()
-        state_ids = set()
+    def _get_shipping_country(self, values):
+        countries = self.env['res.country']
+        states = self.env['res.country.state']
         values['shipping_countries'] = values['countries']
         values['shipping_states'] = values['states']
 
-        DeliveryCarrier = self.pool['delivery.carrier']
-        delivery_carrier_ids = DeliveryCarrier.search(cr, SUPERUSER_ID, [('website_published', '=', True)], context=context)
-        for carrier in DeliveryCarrier.browse(cr, SUPERUSER_ID, delivery_carrier_ids, context):
+        delivery_carriers = self.env['delivery.carrier'].search([('website_published', '=', True)])
+        for carrier in delivery_carriers:
             if not carrier.country_ids and not carrier.state_ids:
                 return values
             # Authorized shipping countries
-            country_ids = country_ids|set(carrier.country_ids.ids)
+            countries = countries | carrier.country_ids
             # Authorized shipping countries without any state restriction
-            state_country_ids = [country.id for country in carrier.country_ids if country.id not in carrier.state_ids.mapped('country_id.id')]
+            state_countries = carrier.country_ids.filtered(lambda country: country.id not in carrier.state_ids.mapped('country_id.id'))
             # Authorized shipping states + all states from shipping countries without any state restriction
-            state_ids = state_ids|set(carrier.state_ids.ids)|set(values['states'].filtered(lambda r: r.country_id.id in state_country_ids).ids)
+            states = states | carrier.state_ids | values['states'].filtered(lambda state: state.country_id.id in state_countries.ids)
 
-        values['shipping_countries'] = values['countries'].filtered(lambda r: r.id in list(country_ids))
-        values['shipping_states'] = values['states'].filtered(lambda r: r.id in list(state_ids))
+        values['shipping_countries'] = values['countries'].filtered(lambda country: country.id in countries.ids)
+        values['shipping_states'] = values['states'].filtered(lambda state: state.id in states.ids)
```

Comment:
Use set intersections here:
```python
values['countries'] & countries
```
will return a recordset of the countries in both values['countries'] and countries, same for states